src/rviz/visualization_frame.cpp

```diff
@@ -1022,12 +1022,67 @@ void VisualizationFrame::addTool( Tool* tool )
   remove_tool_menu_->addAction( tool->getName() );
 }
 
+uint VisualizationFrame::toKey( QString const & str )
+{
+    QKeySequence seq( str );
+    uint keyCode;
+
+    // We should only working with a single key here
+    if( seq.count() == 1 )
+        keyCode = seq[0];
+    else {
+        // Should be here only if a modifier key (e.g. Ctrl, Alt) is pressed.
+        assert( seq.count() == 0 );
+
+        // Add a non-modifier key "A" to the picture because QKeySequence
+        // seems to need that to acknowledge the modifier. We know that A has
+        // a keyCode of 65 (or 0x41 in hex)
+        seq = QKeySequence( str + "+A" );
+        assert( seq.count() == 1);
+        assert( seq[0] > 65 );
+        keyCode = seq[0] - 65;
+    }
+
+    return keyCode;
+}
+
 void VisualizationFrame::onToolbarActionTriggered( QAction* action )
 {
+  // get both the triggered tool and the currently active tool
   Tool* tool = action_to_tool_map_[ action ];
+  Tool* current_tool = manager_->getToolManager()->getCurrentTool();
+
   if( tool )
   {
-    manager_->getToolManager()->setCurrentTool( tool );
+    // if the incoming tool is different from the current one
+    if( tool != current_tool )
+    {
+      // check if the current tool want's to keep control
+      if( current_tool->keepControl() )
+      {
+        // if yes, this means that the incoming shortkey has to be passed to the tool manager
+        QKeyEvent* key = new QKeyEvent( QEvent::KeyPress , toKey( action->shortcut() ), Qt::NoModifier,  action->shortcut() );
```

Comment:
Can you put some of the arguments on new lines to keep the overall line length down? I was imagining something like this:

```c++
QKeyEvent* key = new QKeyEvent( QEvent::KeyPress,
                                toKey( action->shortcut() ),
                                Qt::NoModifier,
                                action->shortcut() );
```

Also make sure to avoid spaces between arguments and the following `,` after then and only use one space between a `,` and the following argument (both were not the case in this line).